editor/store/selectors.js

```diff
@@ -1539,7 +1575,7 @@ export const getInserterItems = createSelector(
 
 			let isDisabled = false;
 			if ( ! hasBlockSupport( blockType.name, 'multiple', true ) ) {
-				isDisabled = some( getBlocks( state ), { name: blockType.name } );
+				isDisabled = some( getBlocksUnfolded( state ), { name: blockType.name } );
```

Comment:
Hi @nosolosw, the data structures we use on state are flat maybe we can take advantage of that. In state.editor.present.blocksByClientId we have an object with all the blocks including the nested ones. So we can just search for a block with that name on the list. Unfortunately, I think we have blocks that are previews so to make sure the block exists on the document if we find something we need to check if the UID is present in state.editor.present.blockOrder. This approach would take advantage of our already flat structures and avoid the new to create new arrays just to make a validation. Do you think this would work?

If we are storing blocks that are not part of the document in state.editor.present.blocksByClientId we should at least flag them with something like is preview blocks so we could make this check with just a search in an array. And plugin developers would be able to distinguish between blocks part of the document and blocks that are not part of the document.

Reply:
I wish we could do that, but in my testing, I've found that `editor.present.blockOrder` and `editor.present.blocksByUID` end up containing blocks that are not present in the document, so we can't operate solely on those. The only reliable mechanism I've found is to build the block tree from the `blockOrder['']`.

Reply:
In ae57038 I've modified the external API to expose the `getBlockIdsUnfolded` selectors instead of `getBlocksUnfolded`.

That way the API still has only two ways to query blocks:

- `getBlocks` for top-level blocks
- `getBlocksByUID` for retrieving any kind of blocks by IDs

which is arguably simpler to understand, but still exposes a helper selector to query all blocks, including referenced by shared and children of nested.

Reply:
If we find a block in blocksByUID, can we check if the block is, in fact, present in the document by checking if the block is referenced by some block or the root block in state.editor.present.blockOrder?

If that's not the case I feel plugin developers will also not be able to know what blocks are present in the document. So probably we should add some flags to specify if a block is a preview block or not (in a separate PR).

Reply:
> If we find a block in blocksByUID, can we check if the block is, in fact, present in the document by checking if the block is referenced by some block or the root block in state.editor.present.blockOrder?

Yes, that works. This is what I mean by "building the block tree from the `blockOrder['']`".

Note that if you want to check if a block is referenced by a shared block that wouldn't work because the relationship is encoded through the `ref` attribute in the block contents.

> If that's not the case I feel plugin developers will also not be able to know what blocks are present in the document. So probably we should add some flags to specify if a block is a preview block or not (in a separate PR).

What's a preview block?

Reply:
My idea would be to a check if a block is part of the document, by checking if the block is referenced by any other block in blockOrder or by blockOrder[''] without needing to iterate on the tree just checking the array.
````

forEach( blockOrder as blockUids) {
 if( blockUIs contains blockUid ) {
	 //block uid is part of the document.
 } 
}
```
Do you think that this would work?

> What's a preview block?

I think the blocks present in blocksByUID that are not part of the document are used to display block previews. Maybe we can flag this blocks so just by iterating on blocksByUID we can know what blocks are part of the document or not.
I just discovered a bug related with this problem if we have shared blocks, we create an empty post we preview shared blocks on the inserter without previewing them our block counts are wrong because the counter thinks all blocks in blocksByUID are part of the document.

Reply:
Summing up here some bits that Jorge and I discussed privately: 

- preview blocks are the ones displayed in the inserter when you hover over the shared blocks. See:

<img width="552" alt="screen shot 2018-07-24 at 09 30 25" src="https://user-images.githubusercontent.com/583546/43131419-44b3c0a8-8f39-11e8-924f-4ce302334674.png">

- shared blocks (although not used) are appended to `blocksByClientId`.

- `blocksByClientId` is updated every time any block attribute changes (like in any key press for paragraphs). The cache of any selector that is dependant on it will be invalidated every time it changes, rendering the cache unuseful.

- can the `getBlockIdsUnfolded` selector only depend on `blockOrder`? No. Some plugin may update the `blocksByClientId` by changing the `ref` attribute of a shared block present in the document. That change wouldn't be reflected in `blockOrder`.

- the outstanding blocker for merging this would be performance. We don't know the impact of this for real, but we are operating on this assumption: avoid array concatenation (prefer [flatMap](https://lodash.com/docs/4.17.10#flatMap)).

Reply:
The selector has been refactored to use `flatMap`.